File: backend/app/api/routes/duty_schedules.py
from datetime import date
from typing import Optional

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, require_roles
from app.core.database import get_db
from app.core.roles import CONTENT_ROLES
from app.models.user import User
from app.schemas.duty_schedule import (
    DutyDayBoard,
    DutyScheduleCreate,
    DutyScheduleOut,
    DutyWeekBoard,
)
from app.services import duty_schedule_service

router = APIRouter(
    prefix="/duty-schedules",
    tags=["duty-schedules"],
    dependencies=[Depends(get_current_user)],
)

# Ghi chu: tao dong ca truc di qua bang cha -> POST /duty-week-plans/{id}/entries.
# Router nay chi con doc (danh sach + bang tong hop) va sua/xoa tung dong.


@router.get("", response_model=list[DutyScheduleOut], status_code=status.HTTP_200_OK)
def list_duties(
    skip: int = 0,
    limit: int = 100,
    date_from: Optional[date] = None,
    date_to: Optional[date] = None,
    unit_id: Optional[int] = None,
    duty_type: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return duty_schedule_service.list_duties(
        db, skip, limit, date_from, date_to, unit_id, duty_type, current_user=current_user
    )


@router.get("/board/day", response_model=DutyDayBoard, status_code=status.HTTP_200_OK)
def duty_day_board(
    day: date = Query(..., description="Ngày cần xem kíp trực toàn Lữ đoàn"),
    unit_id: Optional[int] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Tính năng 1: kíp trực toàn Lữ đoàn theo ngày, gom theo đơn vị + tổng quân số.

    Chỉ huy Lữ đoàn thấy mọi trạng thái bảng trực tuần (kèm nhãn); đơn vị cấp
    dưới chỉ thấy bảng đã duyệt của đơn vị khác và mọi trạng thái của đơn vị mình.
    """
    return duty_schedule_service.build_day_board(db, day, current_user, unit_id)


@router.get("/board/week", response_model=DutyWeekBoard, status_code=status.HTTP_200_OK)
def duty_week_board(
    week_of: date = Query(..., description="Ngày bất kỳ trong tuần cần xem (chuẩn hoá về Thứ Hai)"),
    unit_id: Optional[int] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Tính năng 2: trực tuần - khung nhìn 7 ngày (Thứ Hai → Chủ Nhật) + ma trận
    trạng thái phê duyệt bảng trực tuần của từng đơn vị."""
    return duty_schedule_service.build_week_board(db, week_of, current_user, unit_id)


@router.get("/{entry_id}", response_model=DutyScheduleOut, status_code=status.HTTP_200_OK)
def get_duty(
    entry_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return duty_schedule_service.get_duty_or_404(db, entry_id, current_user)


@router.put(
    "/{entry_id}",
    response_model=DutyScheduleOut,
    status_code=status.HTTP_200_OK,
    dependencies=[Depends(require_roles(*CONTENT_ROLES))],
)
def update_duty(
    entry_id: int,
    entry_in: DutyScheduleCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return duty_schedule_service.update_entry(db, entry_id, entry_in, current_user)


@router.delete(
    "/{entry_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(require_roles(*CONTENT_ROLES))],
)
def delete_duty(
    entry_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    duty_schedule_service.delete_entry(db, entry_id, current_user)
    return None

File: frontend/src/api/dutySchedules.ts
import { apiClient } from './client'
import type {
  DutyDayBoard,
  DutyPlanAttachment,
  DutySchedule,
  DutyScheduleCreate,
  DutyWeekBoard,
  DutyWeekPlan,
  DutyWeekPlanCreate,
  DutyWeekPlanDetail,
  DutyWeekPlanReview,
  DutyWeekPlanUpdate,
} from '../types/dutySchedule'

/** Dong ca truc: chi con doc danh sach + sua/xoa tung dong + 2 bang tong hop. */
export const dutySchedulesApi = {
  list: (params?: {
    skip?: number
    limit?: number
    dateFrom?: string
    dateTo?: string
    unitId?: number
    dutyType?: string
  }) => {
    const q = new URLSearchParams()
    if (params?.skip !== undefined) q.set('skip', String(params.skip))
    if (params?.limit !== undefined) q.set('limit', String(params.limit))
    if (params?.dateFrom) q.set('date_from', params.dateFrom)
    if (params?.dateTo) q.set('date_to', params.dateTo)
    if (params?.unitId !== undefined) q.set('unit_id', String(params.unitId))
    if (params?.dutyType) q.set('duty_type', params.dutyType)
    const qs = q.toString()
    return apiClient.get<DutySchedule[]>(`/duty-schedules${qs ? `?${qs}` : ''}`)
  },
  get: (id: number) => apiClient.get<DutySchedule>(`/duty-schedules/${id}`),
  updateEntry: (id: number, entry: DutyScheduleCreate) =>
    apiClient.put<DutySchedule>(`/duty-schedules/${id}`, entry),
  removeEntry: (id: number) => apiClient.delete<void>(`/duty-schedules/${id}`),

  /** Tinh nang 1: kip truc toan Lu doan theo ngay, gom theo don vi. */
  dayBoard: (day: string, unitId?: number) => {
    const q = new URLSearchParams({ day })
    if (unitId !== undefined) q.set('unit_id', String(unitId))
    return apiClient.get<DutyDayBoard>(`/duty-schedules/board/day?${q.toString()}`)
  },
  /** Tinh nang 2: truc tuan - khung nhin 7 ngay + ma tran phe duyet theo don vi. */
  weekBoard: (weekOf: string, unitId?: number) => {
    const q = new URLSearchParams({ week_of: weekOf })
    if (unitId !== undefined) q.set('unit_id', String(unitId))
    return apiClient.get<DutyWeekBoard>(`/duty-schedules/board/week?${q.toString()}`)
  },
}

/** Bang truc tuan cua don vi: lap - trinh - chi huy phe duyet. */
export const dutyWeekPlansApi = {
  list: (params?: { unitId?: number; weekOf?: string; status?: string }) => {
    const q = new URLSearchParams()
    if (params?.unitId !== undefined) q.set('unit_id', String(params.unitId))
    if (params?.weekOf) q.set('week_of', params.weekOf)
    if (params?.status) q.set('status_filter', params.status)
    const qs = q.toString()
    return apiClient.get<DutyWeekPlan[]>(`/duty-week-plans${qs ? `?${qs}` : ''}`)
  },
  get: (id: number) => apiClient.get<DutyWeekPlanDetail>(`/duty-week-plans/${id}`),
  create: (plan: DutyWeekPlanCreate) =>
    apiClient.post<DutyWeekPlan>('/duty-week-plans', plan),
  update: (id: number, data: DutyWeekPlanUpdate) =>
    apiClient.put<DutyWeekPlan>(`/duty-week-plans/${id}`, data),
  remove: (id: number) => apiClient.delete<void>(`/duty-week-plans/${id}`),
  addEntry: (id: number, entry: DutyScheduleCreate) =>
    apiClient.post<DutySchedule>(`/duty-week-plans/${id}/entries`, entry),
  submit: (id: number) => apiClient.post<DutyWeekPlan>(`/duty-week-plans/${id}/submit`, {}),
  review: (id: number, review: DutyWeekPlanReview) =>
    apiClient.post<DutyWeekPlan>(`/duty-week-plans/${id}/review`, review),
  reopen: (id: number) => apiClient.post<DutyWeekPlan>(`/duty-week-plans/${id}/reopen`, {}),

  // --- Tep dinh kem cua bang truc tuan (nhieu tep moi loai / bang) ---
  listAttachments: (planId: number) =>
    apiClient.get<DutyPlanAttachment[]>(`/duty-week-plans/${planId}/attachments`),
  uploadAttachment: (planId: number, file: File, label?: string) => {
    const fd = new FormData()
    fd.append('file', file)
    if (label?.trim()) fd.append('label', label.trim())
    return apiClient.postForm<DutyPlanAttachment>(`/duty-week-plans/${planId}/attachments`, fd, {
      successMessage: 'Đã đính kèm tệp lịch trực',
    })
  },
  deleteAttachment: (planId: number, attachmentId: number) =>
    apiClient.delete<void>(`/duty-week-plans/${planId}/attachments/${attachmentId}`),
}
